src/com/amazon/ion/impl/lite/IonSystemLite.java

```diff
@@ -317,156 +317,19 @@ public SymbolTable newSharedSymbolTable(String name,
-            String text = token.getText();
-            if (text != null && token.getSid() != UNKNOWN_SYMBOL_ID)
-            {
-                token = newSymbolToken(text, UNKNOWN_SYMBOL_ID);
-            }
-            v.setFieldNameSymbol(token);
-            symbol_is_present = true;
-        }
-
-        // Forget any incoming SIDs on annotations.
-        // This is a fresh array so we can modify it:
-        SymbolToken[] annotations = reader.getTypeAnnotationSymbols();
-        if (annotations.length != 0)
-        {
-            for (int i = 0; i < annotations.length; i++)
-            {
-                SymbolToken token = annotations[i];
-                String text = token.getText();
-                if (text != null && token.getSid() != UNKNOWN_SYMBOL_ID )
-                {
-                    annotations[i] = newSymbolToken(text, UNKNOWN_SYMBOL_ID);
-                }
-            }
-            v.setTypeAnnotationSymbols(annotations);
-            symbol_is_present = true;
-        }
-
-        if (!reader.isNullValue()) {
-            switch (t) {
-            case BOOL:
-            case INT:
-            case FLOAT:
-            case DECIMAL:
-            case TIMESTAMP:
-            case SYMBOL:
-            case STRING:
-            case CLOB:
-            case BLOB:
-                break;
-            case LIST:
-            case SEXP:
-            case STRUCT:
-                // we have to load the children after we grabbed the
-                // fieldname and annotations off of the parent container
-                if (load_children((IonContainerLite)v, reader)) {
-                    symbol_is_present = true;
-                }
-                break;
-            default:
-                throw new IonException("unexpected type encountered reading value: "+t.toString());
-            }
-        }
-        if (symbol_is_present) {
-            v._isSymbolPresent(true);
-        }
-        return v;
-    }
-
-    /**
-     * @return true iff any child contains a symbol
-     * (including field names and annotations)
-     */
-    private boolean load_children(IonContainerLite container, IonReader reader)
-    {
-        boolean symbol_is_present = false;
-
-        reader.stepIn();
-        for (;;) {
-            IonType t = reader.next();
-            if (t == null) {
-                break;
-            }
-            IonValueLite child = load_value_helper(reader, /*isTopLevel*/ false);
-
-            container.add(child);
-
-            if (child._isSymbolPresent()) {
-                symbol_is_present = true;
-            }
-        }
-        reader.stepOut();
-
-        return symbol_is_present;
+        // Note: this method constructs a new `ValueLoader` on each call to preserve thread safety.
+        // If this causes excessive GC pressure, we should consider making a thread-local ValueLoader member field
+        // on the IonSystemLite class.
+        return new ValueLoader().load(reader);
```

Comment:
🗺️ `ValueLoader` (defined below) is a new helper class that encapsulates the logic for materializing values iteratively instead of recursively.